### PR TITLE
purge duplicated bank prioritization fee from cache

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

Duplicated banks can add inconsistent data to prioritization_fee_cache.

#### Summary of Changes
- purge duplicated bank's priority_fee from cache during finalization
- add test for the described scenario
- add metrics counts to monitor such occurrences 

Fixes #

### CODE REVIEW

File: rpc/src/optimistically_confirmed_bank_tracker.rs [hunk 1]
```diff
@@ -296,6 +297,9 @@ impl OptimisticallyConfirmedBankTracker {
                         *highest_confirmed_slot = slot;
                     }
                     drop(w_optimistically_confirmed_bank);
+
+                    // finalize block's minimum prioritization fee cache for this bank
+                    prioritization_fee_cache.finalize_priority_fee(slot, bank_id);
```
Comment (taozhu-chicago) [inline rpc/src/optimistically_confirmed_bank_tracker.rs:302]: finalize `prioritization_fee_cache` with OC-ed bank_id, purge fee from other banks from cache.

File: rpc/src/optimistically_confirmed_bank_tracker.rs [hunk 2]
```diff
@@ -356,6 +358,9 @@ impl OptimisticallyConfirmedBankTracker {
                         w_optimistically_confirmed_bank.bank = bank;
                     }
                     drop(w_optimistically_confirmed_bank);
+
+                    // finalize block's minimum prioritization fee cache for this bank
+                    prioritization_fee_cache.finalize_priority_fee(frozen_slot, frozen_bank_id);
```
Comment (taozhu-chicago) [inline rpc/src/optimistically_confirmed_bank_tracker.rs:363]: do same when pending OC bank is frozen

File: runtime/src/prioritization_fee_cache.rs [hunk 3]
```diff
@@ -121,20 +109,26 @@ impl PrioritizationFeeCacheMetrics {
 enum CacheServiceUpdate {
     TransactionUpdate {
         slot: Slot,
+        bank_id: BankId,
```
Comment (taozhu-chicago) [inline runtime/src/prioritization_fee_cache.rs:112]: add `bank_id` to `CacheServioceUpdate` enums

File: rpc/src/optimistically_confirmed_bank_tracker.rs (hunk 1 above)
Comment (ryoqun) [reply rpc/src/optimistically_confirmed_bank_tracker.rs:302]: hmm, actually, i think `.finalize_priority_fee()` should be called just after `Self::notify_slot_status(..., SlotNotification::OptimisticallyConfirmed)`-ing inside `OptimisticallyConfirmedBankTracker::notify_or_defer()`.

otherwise, we can't ensure bank is frozen (prioritization cache can potentially have updates until frozen) and all parent banks' prioritization fee caches are correctly finalized without skipping?

note that `BankNotification::OptimisticallyConfirmed(slot)` can have gaps and could be send before locally replaying..

so, #32692 had already bugs, it seems...

File: runtime/src/prioritization_fee_cache.rs [hunk 4]
```diff
@@ -121,20 +109,26 @@ impl PrioritizationFeeCacheMetrics {
...
         transaction_fee: u64,
         writable_accounts: Arc<Vec<Pubkey>>,
     },
     BankFrozen {
         slot: Slot,
+        bank_id: BankId,
     },
     Exit,
```
Comment (ryoqun) [inline runtime/src/prioritization_fee_cache.rs:120]: hmm, how about adding new `BankDumped { slot: Slot }`, instead of introducing this `BankId`-based 2nd level caching? And the variant can be sent inside `purge_unfirmed_duplicate_slot()` after plumbing `&PriotizationFeeCache` there.

that alternative impl is simpler and should be safe as long as `solPrFeeCacSvc` is single threaded, which should be because of already existing linearizable  requirement of `BankFrozen`.

Comment (taozhu-chicago) [reply runtime/src/prioritization_fee_cache.rs:120]: Agreed. But then I thought that if I had missed `purge_unconfirmed_duplicated_slot()` last time, what would make sure I won't miss something else, or new duplicated slot/bank logic being added without being aware of `prioritization_fee_cache`? That motivates to explore the approach of purge cache by OC event, which should be definitive and un-rollbackable. But I do agree it adds some complicity, and using `bank_id` as secondary key may not be ideal. Hence this is a draft. 😃  Keen to hear your thoughts on it.

File: rpc/src/optimistically_confirmed_bank_tracker.rs (hunk 1 above)
Comment (taozhu-chicago) [reply rpc/src/optimistically_confirmed_bank_tracker.rs:302]: Mmm, not familiar enough with this part of code. Let me understand how it flows first.

File: runtime/src/prioritization_fee_cache.rs (hunk 4 above)
Comment (ryoqun) [reply runtime/src/prioritization_fee_cache.rs:123]: yeah, i see your reasoning for bank_id approach clearly.

i spent some time on which is better. and i couldn't conclude it. so, it boils down to the author's preference.

as said bank_id takes more human effort than `BankDumped` approach. yet, i saw you added tests. seems you're well energized for the effort. :)

please continue. once this pr is out of draft, i'll review.

File: rpc/src/optimistically_confirmed_bank_tracker.rs (hunk 1 above)
Comment (taozhu-chicago) [reply rpc/src/optimistically_confirmed_bank_tracker.rs:302]: yea, the gap between OptimisticallyConfirmed and bank frozen is real. Thanks for pointing to the correct solution. I opened a separate PR #33100 to address this bug.

File: runtime/src/prioritization_fee_cache.rs (hunk 4 above)
Comment (taozhu-chicago) [reply runtime/src/prioritization_fee_cache.rs:123]: Totally agree that this approach is bit more involved, but not too bad :) I reasoned to myself that it is worth the extra effort to make p.fee cache bit more independent.

I added test thinking either approach we will need a test to capture the corner case scenario. 

I'd like to add a metric counter to capture the occurrence of dup bank, then will move it out of draft.

Thanks for being patient ❤️

File: runtime/src/prioritization_fee_cache.rs [hunk 5]
```diff
@@ -121,20 +125,26 @@ impl PrioritizationFeeCacheMetrics {
 enum CacheServiceUpdate {
     TransactionUpdate {
         slot: Slot,
+        bank_id: BankId,
         transaction_fee: u64,
         writable_accounts: Arc<Vec<Pubkey>>,
     },
     BankFrozen {
```
Comment (ryoqun) [inline runtime/src/prioritization_fee_cache.rs:132]: come to think of it, this variant isn't about `frozen`. maybe, `BankFinalized`?

**Review by ryoqun — COMMENTED** [review]
generally looks good. i think #33100 needs to land first and rebase this pr then.

File: runtime/src/prioritization_fee_cache.rs (hunk 5 above)
Comment (taozhu-chicago) [reply runtime/src/prioritization_fee_cache.rs:132]: Oh, right! fixing it.

**Comment by taozhu-chicago** [comment]
> generally looks good. i think #33100 needs to land first and rebase this pr then.

Yea, this sounds a plan

**Comment by taozhu-chicago** [comment]
> > generally looks good. i think #33100 needs to land first and rebase this pr then.
> 
> Yea, this sounds a plan

#33100 is merged and rebased.

**Review by apfitzge — COMMENTED** [review]
Don't see any issues, but I think the bug was found by ryoqun, so please wait for his feedback before merging

File: runtime/src/prioritization_fee_cache.rs [hunk 6]
```diff
@@ -368,21 +400,24 @@ impl PrioritizationFeeCache {
...
-                        {
-                            fee = std::cmp::max(fee, account_fee);
-                        }
-                    }
-                    Some((*slot, fee))
-                })
+            .filter_map(|(slot, slot_prioritization_fee)| {
+                slot_prioritization_fee
+                    .iter()
+                    .find_map(|prioritization_fee_read| {
```
Comment (ryoqun) [inline runtime/src/prioritization_fee_cache.rs:406]: nit: how about renaming `prioritization_fee_read` => `prioritization_fee` as this isn't related to `Mutex` anymore.

File: runtime/src/prioritization_fee_cache.rs [hunk 7]
```diff
@@ -278,71 +289,88 @@ impl PrioritizationFeeCache {
...
+        let (slot_prioritization_fee, cache_lock_time) =
             measure!(Self::get_prioritization_fee(cache, slot), "cache_lock_time");
 
-        let (mut block_prioritization_fee, entry_lock_time) =
-            measure!(block_prioritization_fee.lock().unwrap(), "entry_lock_time");
-
         // prune cache by evicting write account entry from prioritization fee if its fee is less
         // or equal to block's minimum transaction fee, because they are irrelevant in calculating
         // block minimum fee.
         let (_, slot_finalize_time) = measure!(
```
Comment (ryoqun) [inline runtime/src/prioritization_fee_cache.rs:327]: nit: seems the returned Result is silently discarded after assigning it to `_`. how about `error!()`-ing, at least?

side-note: i have hidden desire for `--deny clippy::let_underscore_must_use` tree-wide... ;)

File: runtime/src/prioritization_fee_cache.rs [hunk 8]
```diff
@@ -121,20 +125,26 @@ impl PrioritizationFeeCacheMetrics {
 enum CacheServiceUpdate {
     TransactionUpdate {
         slot: Slot,
+        bank_id: BankId,
         transaction_fee: u64,
         writable_accounts: Arc<Vec<Pubkey>>,
     },
-    BankFrozen {
+    BankFinalized {
```
Comment (ryoqun) [inline runtime/src/prioritization_fee_cache.rs:132]: :heart:

File: runtime/src/prioritization_fee_cache.rs [hunk 9]
```diff
@@ -438,9 +473,8 @@ mod tests {
 
         // wait till update is done
         while block_fee
-            .lock()
-            .unwrap()
-            .get_min_transaction_fee()
+            .get(&bank.bank_id())
+            .map(|block_fee| block_fee.get_min_transaction_fee())
```
Comment (ryoqun) [inline runtime/src/prioritization_fee_cache.rs:477]: calling `block_fee.get_min_transaction_fee()` isn't meaningless... i think you meant to call `.and_then()`? or remove `.map(...)` altogether?

Comment (ryoqun) [reply runtime/src/prioritization_fee_cache.rs:477]: i mean, i think `.map(...)`'s return type is `Option<Option<u64>>`. so, we aren't checking `.is_none()` on the returned value of `.get_min_transaction_fee()`.

File: runtime/src/prioritization_fee_cache.rs [hunk 10]
```diff
@@ -451,15 +485,19 @@ mod tests {
...
         let fee = PrioritizationFeeCache::get_prioritization_fee(
             prioritization_fee_cache.cache.clone(),
             &slot,
         );
 
         // wait till finalization is done
-        while !fee.lock().unwrap().is_finalized() {
+        while !fee
+            .get(&bank_id)
+            .map_or_else(|| false, |block_fee| block_fee.is_finalized())
```
Comment (ryoqun) [inline runtime/src/prioritization_fee_cache.rs:499]: nit: `.map_or()` should be preferred here, because it's fine to eager-evaluate `false`.

**Review by ryoqun — APPROVED** [review]
lgtm with nits

File: runtime/src/prioritization_fee_cache.rs (hunk 7 above)
Comment (taozhu-chicago) [reply runtime/src/prioritization_fee_cache.rs:327]: 👍🏼  added `error!(...)` should Result is_err()

File: runtime/src/prioritization_fee_cache.rs (hunk 9 above)
Comment (taozhu-chicago) [reply runtime/src/prioritization_fee_cache.rs:477]: Good catch! Yea, should use `and_then`; then realized just checking if `block_fee.get_min_transaction_fee()` has value would make test flaky. In case a `bank` updates cache with multiple transactions, since each transaction applies to cache one at time, the `sync_update()` should make sure all transactions have applied (instead of checking if _any_ transaction has applied). I moved `metrics.accumulate_successful_transaction_update_count(1);` and _conveniently_ used it to sync updates. tbh, not too happy double-purposing metrics counter here.

Comment (ryoqun) [reply runtime/src/prioritization_fee_cache.rs]: oh, glad that we fixed test flakiness. :)

> tbh, not too happy double-purposing metrics counter here.

same here. but i think this is acceptable, considering this is just a tiny piece of test harnesses...

**Review by ryoqun — APPROVED** [review]
lgtm. thanks for the extra polishes. :)

**Comment by taozhu-chicago** [comment]
will added bp label when #33100 bps are merged.
